### PR TITLE
change_hooks

### PR DESCRIPTION
**Pull request by tardyp**
I can see there are a lot of change hooks PR right now.

As they are not yet integrated in nine, we cannot say: nine-first! for them.

So here is the integration PR.

nothing fancy:
* bunch of git mv
* bunch of autopep8
* copy paste from eight's baseweb.py for integration

I'm not so found of the module+functions based design for this, but hey this is working.

### CODE REVIEW

**Comment by sa2ajj** [comment]
Tests:

```
FAILED (skips=34, errors=34, successes=4491)
```

**Comment by tardyp** [comment]
oops

**Comment by tardyp** [comment]
works with buildbot_travis:
foo

File: master/buildbot/www/hooks/gitorious.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    changes = []
+    newrev = payload['after']
+
+    branch = payload['ref']
+    if re.match(r"^0*$", newrev):
+        log.msg("Branch `%s' deleted, ignoring" % branch)
+        return []
+    else:
+        for commit in payload['commits']:
+            files = []
```
Comment (yetanotherion) [inline master/buildbot/www/hooks/gitorious.py]: why not putting
'files': []
l.66 ?

File: master/buildbot/www/hooks/gitorious.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+            # Gitorious doesn't send these, maybe later
+            # if 'added' in commit:
+            #     files.extend(commit['added'])
+            # if 'modified' in commit:
+            #     files.extend(commit['modified'])
+            # if 'removed' in commit:
+            #     files.extend(commit['removed'])
+            when_timestamp = dateparse(commit['timestamp'])
+
+            log.msg("New revision: %s" % commit['id'][:8])
```
Comment (yetanotherion) [inline master/buildbot/www/hooks/gitorious.py]: % (commit['id'][:8],)
or
.format

File: master/buildbot/www/hooks/gitorious.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    return (changes, 'git')
+
+
+def process_change(payload, user, repo, repo_url, project):
+    changes = []
+    newrev = payload['after']
+
+    branch = payload['ref']
+    if re.match(r"^0*$", newrev):
+        log.msg("Branch `%s' deleted, ignoring" % branch)
```
Comment (yetanotherion) [inline master/buildbot/www/hooks/gitorious.py]: % (branch,)
or 
.format

File: master/buildbot/www/hooks/base.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    args = request.args
+
+    # first, convert files, links and properties
+    files = None
+    if args.get('files'):
+        files = json.loads(args.get('files')[0])
+    else:
+        files = []
+
+    properties = None
```
Comment (yetanotherion) [inline master/buildbot/www/hooks/base.py]: properties = json.loads(args.get('properties', ['{}'])[0])
might work

File: master/buildbot/www/hooks/base.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        """
+        if (isinstance(value, type([]))):
+            return value[0]
+        else:
+            return value
+
+    args = request.args
+
+    # first, convert files, links and properties
+    files = None
```
Comment (yetanotherion) [inline master/buildbot/www/hooks/base.py]: same here
files = json.loads(args.get('files', ['[]'])[0])

File: master/buildbot/www/hooks/gitorious.py (hunk 1 above)
Comment (tardyp) [inline master/buildbot/www/hooks/gitorious.py]: I did not look at the actual implementation style of the hooks. This is not the purpose of this PR

File: master/buildbot/www/hooks/gitorious.py (hunk 2 above)
Comment (tardyp) [inline master/buildbot/www/hooks/gitorious.py]: idem

File: master/buildbot/www/hooks/gitorious.py (hunk 3 above)
Comment (tardyp) [inline master/buildbot/www/hooks/gitorious.py]: idem

**Comment by yetanotherion** [comment]
I only have minor comments
:+1:  for me

File: master/buildbot/www/hooks/base.py (hunk 5 above)
Comment (tardyp) [inline master/buildbot/www/hooks/base.py]: idem

**Comment by sa2ajj** [comment]
I have somwhat mixed feelings: the codeis take under `www` hierarchy, while the tests stay under `buildbot` one...

I'd prefer that test cases are moved together with the code.

**Comment by tardyp** [comment]
Please have a second look. This is the www hierarchy of buildbot!

Le lun. 11 mai 2015 19:55, Mikhail Sobolev <notifications@github.com> a
écrit :

**Comment by sa2ajj** [comment]
What I mean is that the test cases are in `master/buildbot/test` not in `www/test`.

Yes, I noticed that the test cases are `master/buildbot/test/unit/test_www_*`, but it's still `master` (or, as I put it, `buildbot`) hierarchy.

**Comment by tardyp** [comment]
@sa2ajj 
There are two 'www' directories (relative to git dir):

/www contains the frontend code with small untested python layer to ling to backend via plugins

/master/buildbot/www contains the backend services, which are fully tested in:

/master/buildbot/test/unit like all other buildbot modules.

I agree this looks confusing, we can call /www /www-frontend

**Comment by sa2ajj** [comment]
Right, I should have downloaded the PR earlier and check the names: GitHub strips the important part for such long file names...

**Pull request merged**